### ISSUE TITLE
RL from textual feedback

### ISSUE THREAD

**Issue by zhujiem** [reczoo/reczoo.github.io#1]
+ [Arxiv'26] Reinforcement Learning via Self-Distillation
  > GRPO只能采用scalar reward，信号很稀疏，浪费了一些高质量的textual feedback，比如跑的test cases或错误信息。本文提出SDPO，基于自蒸馏策略优化，即通过输入prompt+feedback作为teacher，只输入prompt作为student，跑OPD loss。其中feedback可以是自己生成的正确答案及当前的执行错误。不同之处是SDPO需要考虑对teacher进行EMA/inital加权，避免训练不稳定(漂移)。
+ Text2grad: Reinforcement learning from natural language feedback. 
+ Optimizing generative ai by backpropagating language model feedback
+ From words to rewards: Leveraging natural language for reinforcement learning
+ Feedback descent: Open-ended text optimization via pairwise comparison
+ Language models can learn from verbal feedback without scalar rewards.

**Comment by zhujiem** [comment]
## Agent自进化

+ Meta-Harness
+ https://github.com/bojieli/ai-agent-book/blob/main/book/chapter9.md

**Comment by zhujiem** [comment]
```
cat fix_dtype.patch 
diff --git a/src/speculators/train/trainer.py b/src/speculators/train/trainer.py
index ee34500..a68088e 100644
--- a/src/speculators/train/trainer.py
+++ b/src/speculators/train/trainer.py
@@ -500,10 +500,10 @@ class Trainer:
             torch.nn.utils.clip_grad_norm_(self.model.parameters(), 1.0)
 
             metrics["error_records_sum"] = torch.tensor(
-                batch["error_records"], dtype=torch.int32, device=loss.device
+                batch["error_records"], dtype=torch.float32, device=loss.device
             )
             metrics["error_records_total"] = torch.tensor(
-                1.0 if self.rank == 0 else 0, device=loss.device
+                1.0 if self.rank == 0 else 0, dtype=torch.float32, device=loss.device
             )
 
             timer.mark("bwd")
```

**Comment by zhujiem** [comment]
## Agent Memory

+ Memagent: Reshaping long-context llm with multi-conv rl-based memory agent
